envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -126,10 +131,28 @@ class RBACFilterFactory(
         return Permission.newBuilder().setHeader(methodMatch).build()
     }
 
-    private fun mapClientToPrincipal(client: String): Principal {
+    private fun mapClientToPrincipals(client: String, snapshot: GlobalSnapshot): List<Principal> {
```

Comment:
How about this?
```kotlin
    private fun mapClientToPrincipals(client: String, snapshot: GlobalSnapshot): Principal {
        val sourceIpPrincipals = if (client in incomingPermissionsProperties.sourceIpAuthentication.enabledForServices) {
            snapshot.endpoints.resources()
                    .filterKeys { client == it }.values
                    .flatMap { clusterLoadAssignment ->
                        clusterLoadAssignment.endpointsList.flatMap { lbEndpoints ->
                            lbEndpoints.lbEndpointsList.map { lbEndpoint ->
                                lbEndpoint.endpoint.address
                            }
                        }
                    }.map { address ->
                        Principal.newBuilder().setSourceIp(CidrRange.newBuilder()
                                .setAddressPrefix(address.socketAddress.address)
                                .setPrefixLen(UInt32Value.of(exactIpMask)).build())
                                .build()
                    }
        } else listOf()

        val clientIdentityHeaderMatcher = HeaderMatcher.newBuilder()
                .setName(incomingPermissionsProperties.clientIdentityHeader).setExactMatch(client).build()
        val clientIdentityPrincipal = Principal.newBuilder().setHeader(clientIdentityHeaderMatcher).build()
        
        val allPrincipals = sourceIpPrincipals + clientIdentityPrincipal
        val orRulesPrincipalSet = Principal.Set.newBuilder().addAllIds(allPrincipals).build()
        
        return Principal.newBuilder().setOrIds(orRulesPrincipalSet).build()
    }
```
First I made principals not null by eagerly providing default value of empty list:
```kotlin
} else listOf()
```
than I removed superfluous list creation
```kotlin
return principals + Principal.newBuilder().setHeader(clientMatch).build()
```
In the end I changes list of principals to single Principal creating `orRules` from Principal.Set but I'm not sure it will work the same.

Reply:
We decided on 1 principal when source base IP auth is used, so this comment will not be valid after my changes.